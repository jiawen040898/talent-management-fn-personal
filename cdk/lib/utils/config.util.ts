import type { Construct } from 'constructs';

import * as productionEnvVariables from '../configs/production.config';
import * as sandboxEnvVariables from '../configs/sandbox.config';
import * as stagingEnvVariables from '../configs/staging.config';
import { CDKEnvironmentVariables } from '../interfaces';

const getEnvironmentVariables = (
    scope: Construct,
    environment: string,
): CDKEnvironmentVariables => {
    const config = {
        sandbox: sandboxEnvVariables.config(scope),
        staging: stagingEnvVariables.config(scope),
        production: productionEnvVariables.config(scope),
    };

    return config[environment as keyof typeof config];
};

export const configUtil = {
    getEnvironmentVariables,
};
