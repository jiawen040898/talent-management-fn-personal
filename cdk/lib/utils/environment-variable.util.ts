import type { Construct } from 'constructs';

import * as productionEnvVariables from '../environment-variables/production-env-variables';
import * as sandboxEnvVariables from '../environment-variables/sandbox-env-variables';
import * as stagingEnvVariables from '../environment-variables/staging-env-variables';

const getEnvironmentVariables = (scope: Construct, environment: string) => {
    const config = {
        sandbox: sandboxEnvVariables.envVariables(scope),
        staging: stagingEnvVariables.envVariables(scope),
        production: productionEnvVariables.envVariables(scope),
    };

    return config[environment as keyof typeof config];
};

export const environmentVariableUtil = {
    getEnvironmentVariables,
};
