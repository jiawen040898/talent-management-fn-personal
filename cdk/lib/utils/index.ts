export * from './config.util';
export * from './esbuild/esbuild-bundling.util';
