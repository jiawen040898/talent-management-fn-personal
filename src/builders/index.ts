export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
