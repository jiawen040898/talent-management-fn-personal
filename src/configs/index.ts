export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
