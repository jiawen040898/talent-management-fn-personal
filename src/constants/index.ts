export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
