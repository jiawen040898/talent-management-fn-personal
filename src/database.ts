import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
