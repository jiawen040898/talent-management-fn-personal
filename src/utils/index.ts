export * from './company.util';
export * from './date.util';
export * from './number.util';
export * from './path.util';
export * from './url-generator.util';
