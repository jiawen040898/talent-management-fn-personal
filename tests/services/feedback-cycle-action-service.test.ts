import { dateTimeUtil, snsService } from '@pulsifi/fn';
import { DataSource, In } from 'typeorm';

import * as database from '../../src/database';
import {
    FeedbackCycle,
    FeedbackCycleReviewee,
    FeedbackCycleReviewer,
    FeedbackCycleStatus,
    RevieweeStatus,
    ReviewerDirection,
    ReviewerStatus,
} from '../../src/models';
import { feedbackCycleActionService } from '../../src/services/feedback-cycle-action.service';
import { dateUtil } from '../../src/utils';
import { getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/test-data';

jest.mock('@pulsifi/fn/services/sns.service');
describe('FeedbackCycleActionService', () => {
    let dataSource: DataSource;
    const spygetDataSource = jest.spyOn(database, 'getDataSource');
    const urlMatchedRegex = '^https://local-employee.pulsifi.me/';
    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesFeedbackCycleActionToBeAdded,
        );
        spygetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('approve manager nomination with reviewee ready for feedback', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();
        jest.mocked(snsService.sendBatchEventModels).mockResolvedValueOnce();

        // Act
        await feedbackCycleActionService.approveManagerNominationAction({
            ...testData.mockCompanyLookupData,
            localDate: dateUtil.getDate(
                dateTimeUtil.addDays(6, new Date()).toISOString(),
            ),
        });

        //Assert
        const result = await dataSource
            .getRepository(FeedbackCycleReviewee)
            .findAndCount({
                select: {
                    id: true,
                    reviewer_selection_manager_id: true,
                },
                where: {
                    feedback_cycle_id:
                        testData.mockFeedbackCycleAutoApprovalData.id,
                    status: RevieweeStatus.READY_FOR_FEEDBACK,
                },
            });
        const totalRevieweeReady = result[1];
        expect(totalRevieweeReady).toBe(2);

        const reviewerResult = await dataSource
            .getRepository(FeedbackCycleReviewer)
            .findAndCount({
                select: {
                    id: true,
                    status: true,
                    reviewer_direction: true,
                },
                where: {
                    feedback_cycle_reviewee_id: In(result[0].map((i) => i.id)),
                    status: ReviewerStatus.READY,
                },
            });
        const totalReviewerReady = reviewerResult[1];
        const totalSelfReviewReady = reviewerResult[0].filter(
            (i) => i.reviewer_direction === ReviewerDirection.SELF,
        ).length;
        expect(totalReviewerReady).toBe(4);

        const uniqueManagerIds = [
            ...new Set(result[0].map((i) => i.reviewer_selection_manager_id)),
        ];

        expect(jest.mocked(snsService.sendEventModel).mock.calls.length).toBe(
            uniqueManagerIds.length,
        );
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                cycle: {
                    id: expect.anything(),
                    start_date: expect.anything(),
                    close_date: expect.anything(),
                    nomination_submission_deadline: expect.anything(),
                    nomination_approval_deadline: expect.anything(),
                    feedback_submission_deadline: expect.anything(),
                },
                manager: {
                    employee_id: expect.anything(),
                },
                receiver: {
                    id: expect.anything(),
                    employee_id: expect.anything(),
                },
                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });

        expect(
            jest.mocked(snsService.sendBatchEventModels).mock.calls.length,
        ).toBe(2);

        expect(
            jest.mocked(snsService.sendBatchEventModels).mock.calls[0][0]
                .length,
        ).toBe(totalRevieweeReady);
        expect(
            jest.mocked(snsService.sendBatchEventModels).mock.calls[1][0]
                .length,
        ).toBe(totalReviewerReady - totalSelfReviewReady);
    });

    it('start cycle with reviewee ready for nomination', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();
        jest.mocked(snsService.sendBatchEventModels).mockResolvedValueOnce();

        // Act
        await feedbackCycleActionService.startCycleWithReadyForNominationAction(
            {
                ...testData.mockCompanyLookupData,
                localDate: dateUtil.getDate(new Date().toISOString()),
            },
        );

        // Assert
        const activeCycleCount = await dataSource
            .getRepository(FeedbackCycle)
            .countBy({
                id: testData.mockFeedbackCycleReadyForNominationData.id,
                status: FeedbackCycleStatus.ACTIVE,
            });
        expect(activeCycleCount).toBe(1);
        const result = await dataSource
            .getRepository(FeedbackCycleReviewee)
            .findAndCount({
                select: {
                    id: true,
                },
                where: {
                    feedback_cycle_id:
                        testData.mockFeedbackCycleReadyForNominationData.id,
                    status: RevieweeStatus.PENDING_NOMINATION,
                },
            });
        const totalPendingNomination = result[1];
        expect(result[1]).toBe(totalPendingNomination);

        const reviewerResult = await dataSource
            .getRepository(FeedbackCycleReviewer)
            .find({
                select: {
                    id: true,
                    status: true,
                    reviewer_direction: true,
                },
                where: {
                    feedback_cycle_reviewee_id: In(result[0].map((i) => i.id)),
                },
            });
        const totalPending = reviewerResult.length;
        const totalSelfAndManagerReady = reviewerResult.filter(
            (i) => i.status === ReviewerStatus.READY,
        ).length;
        expect(totalPending).toBe(4);
        expect(jest.mocked(snsService.sendEventModel).mock.calls.length).toBe(
            totalPendingNomination,
        );
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                cycle: {
                    id: expect.anything(),
                    start_date: expect.anything(),
                    close_date: expect.anything(),
                    nomination_submission_deadline: expect.anything(),
                    nomination_approval_deadline: expect.anything(),
                    feedback_submission_deadline: expect.anything(),
                },
                manager: {
                    employee_id: expect.anything(),
                },
                receiver: {
                    id: expect.anything(),
                    employee_id: expect.anything(),
                },
                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });

        expect(
            jest.mocked(snsService.sendBatchEventModels).mock.calls.length,
        ).toBe(1);
        expect(
            jest.mocked(snsService.sendBatchEventModels).mock.calls[0][0]
                .length,
        ).toBe(totalSelfAndManagerReady);
    });

    it('start cycle with reviewee ready for feedback', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();

        // Act
        await feedbackCycleActionService.startCycleWithReadyForFeedbackAction({
            ...testData.mockCompanyLookupData,
            localDate: dateUtil.getDate(new Date().toISOString()),
        });

        // Assert
        const activeCycleCount = await dataSource
            .getRepository(FeedbackCycle)
            .countBy({
                id: testData.mockFeedbackCycleReadyForFeedbackData.id,
                status: FeedbackCycleStatus.ACTIVE,
            });
        expect(activeCycleCount).toBe(1);
        const result = await dataSource
            .getRepository(FeedbackCycleReviewee)
            .findAndCount({
                select: {
                    id: true,
                },
                where: {
                    feedback_cycle_id:
                        testData.mockFeedbackCycleReadyForFeedbackData.id,
                    status: RevieweeStatus.READY_FOR_FEEDBACK,
                },
            });
        expect(result[1]).toBe(1);

        const readyReviewerCount = await dataSource
            .getRepository(FeedbackCycleReviewer)
            .countBy({
                feedback_cycle_reviewee_id: In(result[0].map((i) => i.id)),
                status: ReviewerStatus.READY,
            });
        expect(readyReviewerCount).toBe(2);

        expect(jest.mocked(snsService.sendEventModel).mock.calls.length).toBe(
            readyReviewerCount,
        );
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                cycle: {
                    id: expect.anything(),
                    start_date: expect.anything(),
                    close_date: expect.anything(),
                    nomination_submission_deadline: expect.anything(),
                    nomination_approval_deadline: expect.anything(),
                    feedback_submission_deadline: expect.anything(),
                },
                provider: {
                    id: expect.anything(),
                    employee_id: expect.anything(),
                },
                receiver: {
                    id: expect.anything(),
                    employee_id: expect.anything(),
                },
                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });
    });
});
