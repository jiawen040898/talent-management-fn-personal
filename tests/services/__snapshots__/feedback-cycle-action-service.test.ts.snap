// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackCycleActionService approve manager nomination with reviewee ready for feedback 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 2",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee9@test.pulsifi.me",
      "name": "First Name 9 Last Name 9",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "manager": {
      "email": "employee9@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 9",
      "full_name": "First Name 9 Last Name 9",
      "last_name": "Last Name 9",
      "user_account_id": 109,
    },
    "receiver": {
      "email": "employee11@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 11",
      "full_name": "First Name 11 Last Name 11",
      "id": Anything,
      "last_name": "Last Name 11",
      "user_account_id": 111,
    },
    "total_nomination": 2,
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_manager_nomination_auto_approved",
  "user_account_id": 109,
}
`;

exports[`FeedbackCycleActionService start cycle with reviewee ready for feedback 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 1",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee6@test.pulsifi.me",
      "name": "First Name 6 Last Name 6",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "email_template": {
      "body_in_html": "<p>Hello, this is a feedback self invite </p>",
      "subject": "You are invite to submit self feedback",
    },
    "provider": {
      "email": "employee7@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 7",
      "full_name": "First Name 7 Last Name 7",
      "id": Anything,
      "last_name": "Last Name 7",
      "user_account_id": 107,
    },
    "receiver": {
      "email": "employee6@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 6",
      "full_name": "First Name 6 Last Name 6",
      "id": Anything,
      "last_name": "Last Name 6",
      "user_account_id": 106,
    },
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_reviewee_response_requested",
  "user_account_id": 107,
}
`;

exports[`FeedbackCycleActionService start cycle with reviewee ready for nomination 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 0",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee0@test.pulsifi.me",
      "name": "First Name 0 Last Name 0",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "manager": {
      "email": "employee5@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 5",
      "full_name": "First Name 5 Last Name 5",
      "last_name": "Last Name 5",
      "user_account_id": 105,
    },
    "receiver": {
      "email": "employee0@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 0",
      "full_name": "First Name 0 Last Name 0",
      "id": Anything,
      "last_name": "Last Name 0",
      "user_account_id": 100,
    },
    "total_nomination": 0,
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_nomination_requested",
  "user_account_id": 100,
}
`;
